### PR TITLE
Added Supplier Invoice Model

### PR DESCRIPTION
**Pull request by make-it-git**
## Overview

| Q                 | A
| ----------------- | ---
| Bugfix?           | no
| New feature?      | yes
| Breaking Changes? | no
| Fixed issues      | #

Notify the following people: @darrynten

### CODE REVIEW

File: tests/SageOne/Models/BaseModelTest.php [hunk 1]
```diff
@@ -968,11 +977,21 @@ protected function setUpRequestMock(
             '{' . $this->config['key'] . '}'
         );
 
+        if (!empty($parameters)) {
+            if ($method === 'GET') {
+                foreach ($parameters as $key => $value) {
+                    $checkParameters['query'][$key] = $value;
+                }
+            } elseif ($method === 'POST' || $method === 'PUT' || $method === 'DELETE') {
```
Comment (darrynten) [inline tests/SageOne/Models/BaseModelTest.php:988]: I don't think the `elseif` is 100% required here

File: src/Models/SupplierInvoice.php [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+     *       "Subject": "sample string 5",
+     *       "Message": "sample string 6",
+     *       "Attachments": [
+     *           {
+     *           "Key": "sample string 1",
+     *           "Value": "sample string 2"
+     *           },
+     *           {
+     *           "Key": "sample string 1",
+     *           "Value": "sample string 2"
```
Comment (darrynten) [inline src/Models/SupplierInvoice.php:354]: Do we know how these files are encoded? Are they sent inline b64 or are there form attachments? Or is it perhaps a reference to a file uploaded in a different mechanism?

File: tests/SageOne/Models/BaseModelTest.php [hunk 3]
```diff
@@ -968,11 +977,21 @@ protected function setUpRequestMock(
...
+                // $checkParameters['json'] = (object)$parameters;
+            }
+        }
+
         /**
         * $client in RequestHandler receives url without query params
         * they are passed as last parameter for $client->request
         */
-        $fullUrl = '//localhost:8082' . $urlWithoutApiKey;
+        $fullUrl = '//localhost:8082' . $urlWithoutParams;
```
Comment (darrynten) [inline tests/SageOne/Models/BaseModelTest.php]: I think we should sprintf this

File: src/Models/SupplierInvoice.php [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+            'max' => 100,
+        ],
+        /**
+         * TODO Find out what is 'printed',
+         * see https://accounting.sageone.co.za/api/1.1.2/Help/Api/GET-SupplierInvoice-Get_includeDetail_includeSupplierDetails
+         * Response has it, but SupplierInvoice does not
+         */
+        'printed' => [
+            'type' => 'boolean',
+            'nullable' => false,
```
Comment (darrynten) [inline src/Models/SupplierInvoice.php]: Should this not be nullable given that it's only in the response? Perhaps even read only too?

**Review by darrynten — CHANGES_REQUESTED** [review]
Nice work @make-it-git some small changes and questions

File: src/Models/SupplierInvoice.php (hunk 2 above)
Comment (make-it-git) [reply src/Models/SupplierInvoice.php:354]: Not yet, once I get to real mocks we can test it

Comment (darrynten) [reply src/Models/SupplierInvoice.php:354]: I think perhaps that we should add a note about that to the docblock and to the Readme

**Comment by make-it-git** [comment]
@darrynten there is one problem
\Mockery:subset() fails for 'json' check because json is passed as stdClass
Even if it is the same object contents it treats them as different objects

http://docs.mockery.io/en/latest/reference/argument_validation.html
When matching objects as arguments, Mockery only does the strict === comparison, which means only the same $object will match:

So I do not know easy way to validate that we passed expected arguments to request in POST/PUT/DELETE methods

**Comment by make-it-git** [comment]
@darrynten any ideas how to test that we passed expected 'json' stdClass?

**Comment by darrynten** [comment]
@make-it-git I sent you a message on slack about it, let's discuss tomorrow morning - this work is looking really good by the way 😎

File: tests/SageOne/Models/BaseModelTest.php [hunk 5]
```diff
@@ -964,15 +977,24 @@ protected function setUpRequestMock(
...
-        );
+
+        $checkParameters['query']['apikey'] = $apikey;
+
+        if (!empty($parameters)) {
+            if ($method === 'GET') {
+                foreach ($parameters as $key => $value) {
+                    $checkParameters['query'][$key] = $value;
+                }
+            } else {
```
Comment (darrynten) [inline tests/SageOne/Models/BaseModelTest.php]: I think we should check that the request is one of the 3 allowed types that can have Json and drop/throw all other verbs

**Review by darrynten — CHANGES_REQUESTED** [review]
I think we need another more robust http verb check

File: tests/SageOne/Models/SupplierInvoiceModelTest.php [hunk 6]
```diff
@@ -0,0 +1,1180 @@
...
+    }
+
+    public function testFeatures()
+    {
+        $this->verifyFeatures(SupplierInvoice::class, [
+            'all' => true, 'get' => true, 'delete' => false, 'save' => true
+        ]);
+    }
+
+    public function testAll()
```
Comment (darrynten) [inline tests/SageOne/Models/SupplierInvoiceModelTest.php:296]: Codacy Issue found: [The method testAll() has 335 lines of code. Current threshold is set to 100. Avoid really long methods.](https://www.codacy.com/app/Xpenso/sage-one-php/pullRequest?prid=804082)

File: tests/SageOne/Models/SupplierInvoiceModelTest.php [hunk 7]
```diff
@@ -0,0 +1,1180 @@
...
+use DarrynTen\SageOne\Models\Supplier;
+use DarrynTen\SageOne\Models\SupplierCategory;
+use DarrynTen\SageOne\Models\TaxType;
+use DarrynTen\SageOne\Models\CommercialDocumentLine;
+use DarrynTen\SageOne\Models\ModelCollection;
+use DarrynTen\SageOne\Exception\ModelException;
+
+class SupplierInvoiceModelTest extends BaseModelTest
+{
+    public function testAttributes()
```
Comment (darrynten) [inline tests/SageOne/Models/SupplierInvoiceModelTest.php:15]: Codacy Issue found: [The method testAttributes() has 273 lines of code. Current threshold is set to 100. Avoid really long methods.](https://www.codacy.com/app/Xpenso/sage-one-php/pullRequest?prid=804082)

File: tests/SageOne/Models/SupplierInvoiceModelTest.php [hunk 8]
```diff
@@ -0,0 +1,1180 @@
...
+
+use DarrynTen\SageOne\Models\SupplierInvoice;
+use DarrynTen\SageOne\Models\Supplier;
+use DarrynTen\SageOne\Models\SupplierCategory;
+use DarrynTen\SageOne\Models\TaxType;
+use DarrynTen\SageOne\Models\CommercialDocumentLine;
+use DarrynTen\SageOne\Models\ModelCollection;
+use DarrynTen\SageOne\Exception\ModelException;
+
+class SupplierInvoiceModelTest extends BaseModelTest
```
Comment (darrynten) [inline tests/SageOne/Models/SupplierInvoiceModelTest.php:13]: Codacy Issue found: [The class SupplierInvoiceModelTest has 1168 lines of code. Current threshold is 1000. Avoid really long classes.](https://www.codacy.com/app/Xpenso/sage-one-php/pullRequest?prid=804082)

File: tests/SageOne/Models/SupplierInvoiceModelTest.php [hunk 9]
```diff
@@ -0,0 +1,1180 @@
...
+        $this->assertEquals(1, $model2->lines->results[1]->analysisCategoryId1);
+        $this->assertEquals(1, $model2->lines->results[1]->analysisCategoryId2);
+        $this->assertEquals(1, $model2->lines->results[1]->analysisCategoryId3);
+        $key = '$trackingCode';
+        $this->assertEquals('sample string 13', $model2->lines->results[1]->$key);
+        $this->assertEquals(1, $model2->lines->results[1]->currencyId);
+        $this->assertEquals(1.0, $model2->lines->results[1]->unitCost);
+    }
+
+    public function testGetId()
```
Comment (darrynten) [inline tests/SageOne/Models/SupplierInvoiceModelTest.php:632]: Codacy Issue found: [The method testGetId() has 166 lines of code. Current threshold is set to 100. Avoid really long methods.](https://www.codacy.com/app/Xpenso/sage-one-php/pullRequest?prid=804082)

File: tests/SageOne/Models/SupplierInvoiceModelTest.php [hunk 10]
```diff
@@ -0,0 +1,1180 @@
...
+        $this->assertEquals(1, $model->lines->results[1]->analysisCategoryId1);
+        $this->assertEquals(1, $model->lines->results[1]->analysisCategoryId2);
+        $this->assertEquals(1, $model->lines->results[1]->analysisCategoryId3);
+        $key = '$trackingCode';
+        $this->assertEquals('sample string 13', $model->lines->results[1]->$key);
+        $this->assertEquals(1, $model->lines->results[1]->currencyId);
+        $this->assertEquals(1.0, $model->lines->results[1]->unitCost);
+    }
+
+    public function testSave()
```
Comment (darrynten) [inline tests/SageOne/Models/SupplierInvoiceModelTest.php:799]: Codacy Issue found: [The method testSave() has 170 lines of code. Current threshold is set to 100. Avoid really long methods.](https://www.codacy.com/app/Xpenso/sage-one-php/pullRequest?prid=804082)

File: tests/SageOne/Models/SupplierInvoiceModelTest.php [hunk 11]
```diff
@@ -0,0 +1,1180 @@
...
+        $this->assertEquals(1, $model->lines->results[1]->analysisCategoryId1);
+        $this->assertEquals(1, $model->lines->results[1]->analysisCategoryId2);
+        $this->assertEquals(1, $model->lines->results[1]->analysisCategoryId3);
+        $key = '$trackingCode';
+        $this->assertEquals('sample string 13', $model->lines->results[1]->$key);
+        $this->assertEquals(1, $model->lines->results[1]->currencyId);
+        $this->assertEquals(1.0, $model->lines->results[1]->unitCost);
+    }
+
+    public function testCalculate()
```
Comment (darrynten) [inline tests/SageOne/Models/SupplierInvoiceModelTest.php:970]: Codacy Issue found: [The method testCalculate() has 172 lines of code. Current threshold is set to 100. Avoid really long methods.](https://www.codacy.com/app/Xpenso/sage-one-php/pullRequest?prid=804082)

File: tests/SageOne/Models/BaseModelTest.php (hunk 5 above)
Comment (darrynten) [inline tests/SageOne/Models/BaseModelTest.php]: Codacy Issue found: [The method verifyGetAll uses an else expression. Else is never necessary and you can simplify the code to work without else.](https://www.codacy.com/app/Xpenso/sage-one-php/pullRequest?prid=804082)

**Comment by darrynten** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 7
           

Complexity increasing per file
==============================
- tests/SageOne/Models/BaseModelTest.php  6
         
```


See the complete overview on [Codacy](https://www.codacy.com/app/Xpenso/sage-one-php/pullRequest?prid=804082&bid=5000561)

**Comment by make-it-git** [comment]
@darrynten Fixed

**Comment by make-it-git** [comment]
@darrynten added better http check

**Review by darrynten — COMMENTED** [review]
Combine a method and then all good

File: tests/SageOne/Models/BaseModelTest.php [hunk 12]
```diff
@@ -922,8 +922,21 @@ protected function setUpRequestMock(
...
         int $responseCode = 200
     ) {
-        $url = sprintf('/1.1.2/%s?apikey=key', $path);
-        $urlWithoutApiKey = sprintf('/1.1.2/%s/', $path);
+        $apikey = urlencode('{key}');
+
+        $url = sprintf('/1.1.2/%s?apikey=%s', $path, $apikey);
+
+        if ($method === 'GET') {
+            if (!empty($parameters)) {
```
Comment (darrynten) [inline tests/SageOne/Models/BaseModelTest.php:930]: I think this line and the one above can be combined here and again further down

File: tests/SageOne/Models/BaseModelTest.php [hunk 13]
```diff
@@ -964,15 +977,29 @@ protected function setUpRequestMock(
                 'Authorization' => sprintf('%s %s', $tokenType, $token)
             ],
         ];
-        $checkParameters['query']['apikey'] = urlencode(
-            '{' . $this->config['key'] . '}'
-        );
+
+        $checkParameters['query']['apikey'] = $apikey;
+
```
Comment (darrynten) [inline tests/SageOne/Models/BaseModelTest.php:982]: Here's the other one. They're actually in a different order to the above one :)

**Comment by darrynten** [comment]
A method call** sorry :)

**Review by darrynten — APPROVED** [review]

**Pull request merged**